action/sealedenvelope.go

```diff
@@ -7,19 +7,39 @@ import (
 	"github.com/iotexproject/iotex-proto/golang/iotextypes"
 	"github.com/pkg/errors"
 
+	"github.com/iotexproject/iotex-core/config"
 	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
 )
 
 // SealedEnvelope is a signed action envelope.
 type SealedEnvelope struct {
 	Envelope
+	encoding      uint32
+	externChainID uint32
+	srcPubkey     crypto.PublicKey
+	signature     []byte
+}
 
-	srcPubkey crypto.PublicKey
-	signature []byte
+// RawHash returns the raw hash of embedded Envelope
+// this is the hash to be signed
+func (sealed *SealedEnvelope) RawHash() hash.Hash256 {
+	if IsRLP(iotextypes.Encoding(sealed.encoding)) {
+		if tx, err := actionToRLP(sealed.Action()); err == nil {
+			return rlpRawHash(tx, sealed.externChainID)
+		}
+		return hash.ZeroHash256
+	}
+	return sealed.Envelope.Hash()
 }
 
 // Hash returns the hash value of SealedEnvelope.
 func (sealed *SealedEnvelope) Hash() hash.Hash256 {
+	if IsRLP(iotextypes.Encoding(sealed.encoding)) {
+		if tx, err := actionToRLP(sealed.Action()); err == nil {
+			return rlpSignedHash(tx, sealed.externChainID, sealed.Signature())
+		}
+		return hash.ZeroHash256
```

Comment:
we should either panic or return the error

Reply:
Hash() returns 32-byte array

i added a check to catch all-0 hash in Verify()

Reply:
If a sealed envelope could be assembled, we shouldn't expect any error here. Thus, we should panic, instead of having some hidden knowledge like ZeroHash

Reply:
yes, ideally, if the signature is correct, rlpSignedHash() will not err.

i feel fallback to 0-hash is safer, b/c 0-hash will fail Verify(), and program will catch the error, and continue. Panic could bring down a node in case a tx somehow caused error in rlpSignedHash() -- though in theory it should not happen